### ISSUE TITLE
Oracle LOB table export is very slow , Need your recommendations to improve performance

### ISSUE THREAD

**Issue by KrishnaMunnangi** [darold/ora2pg#800]
Hi Darold,

We are trying to export around 23 LOB tables( out of which one table is having 7550919 records & other table having 1256893 records, the remaining tables having records starting from 1lakh to 4 laks maximum).

before triggering the ora2pg export we have took below actions on oracle server
 - Increased system resources like CPU cores to 12 & RAM to 32 GB
 - Performed De-Fragmentation where we have fragmented tables
 - Ran Data Dictionary Stats & Gathered tables stats on the target schema.

We have used below ora2pg configuration file to export the LOB tables:

ORACLE_HOME   /oracle/product/rdbms/12.1.0.2
ORACLE_DSN     dbi:Oracle:host=xxxxxxxx;sid=xxxxxxxx;port=xxxxx
ORACLE_USER   xxxxxxxx
ORACLE_PWD    xxxxxxxx
USER_GRANTS     0
EXPORT_SCHEMA   1
CREATE_SCHEMA   1
SCHEMA xxxxxxxxx
PG_SCHEMA xxxxxxxxx
TYPE TABLE PACKAGE INSERT COPY VIEW GRANT SEQUENCE TRIGGER FUNCTION PROCEDURE TABLESPACE TYPE PARTITION FDW MVIEW KETTLE QUERY
ALLOW xxxx xxx xxxx xxxx xxxxx xxxxx xxxxx xxxxx xxxxxx xxxxxx xxxxxx xxxxx xxxxx xxxxxxxx
EXTERNAL_TO_FDW         1
#SKIP   fkeys pkeys ukeys indexes checks
ORA_RESERVED_WORDS      audit,comment,references
USE_RESERVED_WORDS      1
OUTPUT          test.sql
OUTPUT_DIR      /dumplocal/XXXX/LOB_EXPORT
FILE_PER_CONSTRAINT     1
FILE_PER_INDEX          1
FILE_PER_TABLE  1
FILE_PER_FUNCTION      0
PG_NUMERIC_TYPE 1
PG_INTEGER_TYPE 1
DEFAULT_NUMERIC numeric
ENABLE_MICROSECOND      1
DATA_LIMIT      500
BLOB_LIMIT      100
NOESCAPE        0
#TRANSACTION     serializable
TRANSACTION     readonly
STANDARD_CONFORMING_STRINGS     1
LOG_ON_ERROR            0
LONGREADLEN     102400000
TRIM_TYPE               BOTH
INTERNAL_DATE_MAX       49
JOBS            1
ORACLE_COPIES   1
PARALLEL_TABLES 1
DROP_INDEXES    0
SYNCHRONOUS_COMMIT      0
EXPORT_INVALID  0
PLSQL_PGSQL     1
NULL_EQUAL_EMPTY        0
EMPTY_LOB_NULL          0
PACKAGE_AS_SCHEMA       1
REWRITE_OUTER_JOIN      1
DATE_FUNCTION_REWRITE   1
ESTIMATE_COST           1
COST_UNIT_VALUE        10
DUMP_AS_HTML            1
#TOP_MAX                10
HUMAN_DAYS_LIMIT        5
PG_SUPPORTS_WHEN        1
PG_SUPPORTS_INSTEADOF   1
PG_SUPPORTS_MVIEW       1
PG_SUPPORTS_CHECKOPTION 1
PG_SUPPORTS_IFEXISTS    1
BITMAP_AS_GIN           1
PG_BACKGROUND           0
AUTODETECT_SPATIAL_TYPE 1
CONVERT_SRID            1
DEFAULT_SRID            4326
GEOMETRY_EXTRACT_TYPE   INTERNAL
FDW_SERVER      orcl
MYSQL_PIPES_AS_CONCAT           0
MYSQL_INTERNAL_EXTRACT_FORMAT   0
TRUNCATE_TABLE  1
DISABLE_PARTITION 1

the speed of ora2pg you can see below: (on an AVG its exporting 75recs/sec)

nohup ora2pg -J 4 -j 6 -t COPY -o COPY_datas.sql -b /dumplocal/XXXX/LOB_EXPORT_30JUN2019 -c /dumplocal/XXXXX//LOB_XX_ora2pg.conf & 

[======>                  ]  33536/129814 rows (25.8%) Table A-part-2 (994 sec., 33 recs/sec)
[======>                  ]  33645/129814 rows (25.9%) Table A-part-3 (996 sec., 33 recs/sec)
[======>                  ]  33553/129814 rows (25.8%) Table A-part-1 (1000 sec., 33 recs/sec)
[========================>] 519256/519256 rows (100.0%) Table A (1001 sec., 518 recs/sec)
[=====>                   ]  50080/200821 rows (24.9%) Table X-part-2 (4533 sec., 11 recs/sec)
[=====>                   ]  50127/200821 rows (25.0%) Table X-part-0 (4539 sec., 11 recs/sec)
[=====>                   ]  50203/200821 rows (25.0%) Table X-part-3 (4540 sec., 11 recs/sec)
[======>                  ]  50411/200821 rows (25.1%) Table X-part-1 (4550 sec., 11 recs/sec)
[========================>] 803284/803284 rows (100.0%) Table X (4551 sec., 176 recs/sec)
[==========>              ]  5027795/11491314 total rows (43.8%) - (76324 sec., avg: 65 X_A-part-2 in progress.

Could you please suggest any better approach to handle (LOB) and speedup the ora2pg process ?

I am ready to provide you further information if required.

Thanks,
Krishna

**Comment by darold** [comment]
Hi Krishna,

Unfortunately exporting BLOB is slow the entire content need to be translated into hexa. There is no miracle but let's see what could be improved.

What is the value of NO_LOB_LOCATOR? I don't see it in your configuration above. If it is set to 0 you are downloading BLOB by small chunk. Set it to 1.

The value of LONGREADLEN, here 102400000, must correspond to the highest BLOB size, if it is far smallest than that reduce the value as the 100Mb are taken for each row exported (x BLOB_LIMIT). You must have enough RAM to support this allocation especially with multiprocess.

BLOB_LIMIT is a bit too high if you are currently exporting at 65 tuples/sec with 4 Oracle connection, set it to 20 this will free more memory and improve speed on hexa transformation side (x 6 process)

You are using 24 CPU (-J 4 x -j 6) this is the double of the number of CPU you have. You can try first by using ` -J 4 -j 3` to see if it is better, otherwise try the opposite: `-J 2 -j 6`.

Also I see that you have not set DEFINED_PKEY that can be used with `-J` this mean that Ora2Pg will auto detect a Primary or unique key with numeric values from your table. If there is none, Ora2Pg will not user multiprocess to gather data from Oracle, this will result in a `-J 1`.

**Comment by darold** [comment]
Commit 6d887f7 has improved the export speed by 10 times.
``` 
    - Improve speed of BLOB/CLOB data export by more than 10 times.

    Oracle recommends reading from and writing to a LOB in batches using a
    multiple of the LOB chunk size. This chunk size defaults to 8k (8192).
    Recent tests shown that the best performances can be reach with higher
    value like 512K or 4Mb.

    A quick benchmark with 30120 rows with different size of BLOB (200x5Mb,
    19800x212k, 10000x942K, 100x17Mb, 20x156Mb), with DATA_LIMIT=100,
    LONGREADLEN=170Mb and a total table size of 20GB gives:

           no lob locator  : 22m46,218s (1365 sec., avg: 22 recs/sec)
           chunk size 8k   : 15m50,886s (951 sec., avg: 31 recs/sec)
           chunk size 512k : 1m28,161s (88 sec., avg: 342 recs/sec)
           chunk size 4Mb  : 1m23,717s (83 sec., avg: 362 recs/sec)

    In conclusion it can be more than 10 time faster with LOB_CHUNK_SIZE set
    to 4Mb. Depending of the size of most BLOB you may want to adjust the value
    here. For example if you have a majority of small lobs bellow 8K, using 8192
    is better to not waste space. By default LOB_CHUNK_SIZE is set to 512000.
```
